Source/RapyutaSimulationPlugins/Private/Sensors/ROS2CameraComponent.cpp

```diff
@@ -0,0 +1,144 @@
+    // component initialization
+    SceneCaptureComponent = CreateDefaultSubobject<USceneCaptureComponent2D>(TEXT("SceneCaptureComponent"));
+    RenderTarget = CreateDefaultSubobject<UTextureRenderTarget2D>(TEXT("RenderTarget"));
+
+    // Initialize capture and texture component
+    SceneCaptureComponent->SetupAttachment(this);
+}
+
+void UROS2CameraComponent::BeginPlay()
+{
+	Super::BeginPlay();	
+    Init();
+}
+
+void UROS2CameraComponent::Init()
+{
+    SceneCaptureComponent->FOVAngle = FieldOfView;
+    SceneCaptureComponent->OrthoWidth = OrthoWidth;
+    
+    RenderTarget->InitCustomFormat(Width, Height, EPixelFormat::PF_B8G8R8A8, true);
+    SceneCaptureComponent->TextureTarget = RenderTarget;
+
+    // Initialize image data
+    Data.width = Width;
+    Data.height = Height;
+    Data.encoding = Encoding;
+    Data.step = Width * 3; // todo should be variable based on encoding
+    Data.data.AddUninitialized(Width * Height * 3);
+
+    // Node and publisher initialize
+    FActorSpawnParameters SpawnParamsNode;
+    Node = GetWorld()->SpawnActor<AROS2Node>(AROS2Node::StaticClass(), SpawnParamsNode);
+    Node->Name = NodeName.IsEmpty() ? TEXT("UE4CameraNode_" + FGuid::NewGuid().ToString()) : NodeName;
+    Node->Namespace = FString();
+    Node->Init();
+
+    Publisher =  NewObject<UROS2Publisher>(this, UROS2Publisher::StaticClass());
+    Publisher->RegisterComponent();
+    Publisher->TopicName = TopicName;
+    Publisher->PublicationFrequencyHz = PublishFreq;
+    Publisher->MsgClass = UROS2ImageMsg::StaticClass();
+    
+    Publisher->UpdateDelegate.BindDynamic(this, &UROS2CameraComponent::MessageUpdate);
+    Node->AddPublisher(Publisher);
+    Publisher->Init(UROS2QoS::KeepLast);
+}
+
+void UROS2CameraComponent::MessageUpdate(UROS2GenericMsg *TopicMessage)
+{
+    UROS2ImageMsg *Message = Cast<UROS2ImageMsg>(TopicMessage);
+    Message->SetMsg(GetData());
+}
+
+FROSImage UROS2CameraComponent::GetData()
+{
+    SceneCaptureComponent->CaptureScene();
+    CaptureNonBlocking();
+    if(!RenderRequestQueue.IsEmpty()){
+        // Peek the next RenderRequest from queue
+        FRenderRequest* nextRenderRequest = nullptr;
+        RenderRequestQueue.Peek(nextRenderRequest);
+
+        if(nextRenderRequest){ //nullptr check
+            if(nextRenderRequest->RenderFence.IsFenceComplete()){ // Check if rendering is done, indicated by RenderFence
+                for (int I = 0; I < nextRenderRequest->Image.Num(); I++)
+                {
+                    Data.data[I * 3 + 0] = nextRenderRequest->Image[I].R;
+                    Data.data[I * 3 + 1] = nextRenderRequest->Image[I].G;
+                    Data.data[I * 3 + 2] = nextRenderRequest->Image[I].B;
+                    // UE_LOG(LogTemp, Warning, TEXT("AsyncTaskDone"));
+                }
+
+                // Delete the first element from RenderQueue
+                RenderRequestQueue.Pop();
+                delete nextRenderRequest;
+            }
+        }
+    }    

```

Comment:
Since here we do `CaptureNonBlocking()`, which means the GameThread is non-blocking, and the capturing (pixel reading from RHI) could possibly take some time before RHI returns the data, meaning the fence is complete.
Thus, `if(nextRenderRequest->RenderFence.IsFenceComplete())` might not be instantly satisfied.

But more essentially, an instant data fetching function named `GetData()` could be inevitably blocking, if users want to invoke a non-blocking version, they might want to fetch data at a later time, when it is available from RHI, after invoking `CaptureNonBlocking()`?

References:
* TimmHess actually runs the fence check in a [tick](https://github.com/TimmHess/UnrealImageCapture#:~:text=imageWrapper%2D%3EGetCompressed(0)%3B-,RunAsyncImageSaveTask(ImgData%2C%20fileName)%3B,-%7D%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20//%20Delete%20the%20first) then do async image saving.
* [Another one](https://github.com/rapyuta-robotics/RapyutaSimulationInternal/blob/master/Source/RapyutaSimInternal/Private/RapyutaSceneCapture/RRCaptureUtils.cpp#L174) of async fetching style by off-loading the fence check to a thread.

Reply:
Since here we do `CaptureNonBlocking()`, which means the GameThread is non-blocking, and the capturing (pixel reading from RHI) could possibly take some time before RHI returns the data.
Thus, `if(nextRenderRequest->RenderFence.IsFenceComplete())` might not be instantly satisfied.

But more essentially, an instant data fetching function named `GetData()` could be inevitably blocking, if users want to invoke a non-blocking version, they might want to fetch data at a later time, when it is available from RHI, after invoking `CaptureNonBlocking()`?

References:
* TimmHess actually runs the fence check in a [tick](https://github.com/TimmHess/UnrealImageCapture#:~:text=imageWrapper%2D%3EGetCompressed(0)%3B-,RunAsyncImageSaveTask(ImgData%2C%20fileName)%3B,-%7D%0A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20//%20Delete%20the%20first) then do async image saving.
* [Another one](https://github.com/rapyuta-robotics/RapyutaSimulationInternal/blob/master/Source/RapyutaSimInternal/Private/RapyutaSceneCapture/RRCaptureUtils.cpp#L174) of async fetching style by off-loading the fence check to a thread.